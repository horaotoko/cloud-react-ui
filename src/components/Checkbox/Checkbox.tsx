import { forwardRef } from "react";
import styled, { css } from "styled-components";
import { defaultProps } from "../../theme";

export type Props = {
  name?: string;
  defaultChecked?: boolean;
  disabled?: boolean;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick?: (event: React.MouseEvent<HTMLInputElement, MouseEvent>) => void;
  children?: React.ReactNode;
  className?: string;
};

const CheckboxWrapper = styled.span`
  ${({ theme: { checkbox } }) => css`
    position: relative;
    display: inline-block;
    width: ${checkbox.width};
    height: ${checkbox.height};
    vertical-align: middle;
  `}
`;
CheckboxWrapper.defaultProps = defaultProps;

const Input = styled.input`
  ${({ theme: { checkbox } }) => css`
    opacity: 0;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    margin: 0;
    cursor: pointer;

    :checked + ${Box} {
      background-color: ${checkbox.backgroundColor.checked};
      border-width: ${checkbox.borderWidth.checked};
      border-style: ${checkbox.borderStyle.checked};
      border-color: ${checkbox.borderColor.checked};
    }

    :disabled + ${Box} {
      background-color: ${checkbox.backgroundColor.disabled};
      border-width: ${checkbox.borderWidth.disabled};
      border-style: ${checkbox.borderStyle.disabled};
      border-color: ${checkbox.borderColor.disabled};
    }
  `}
`;
Input.defaultProps = defaultProps;

const Box = styled.span`
  ${({ theme: { checkbox } }) => css`
    position: absolute;
    width: 100%;
    height: 100%;
    border-radius: ${checkbox.borderRadius};
    background-color: ${checkbox.backgroundColor.unchecked};
    border-width: ${checkbox.borderWidth.unchecked};
    border-style: ${checkbox.borderStyle.unchecked};
    border-color: ${checkbox.borderColor.unchecked};
    pointer-events: none;
    box-sizing: border-box;
  `}
`;
Box.defaultProps = defaultProps;

const IconWrapper = styled.span`
  ${({ theme: { checkbox } }) => css`
    position: absolute;
    display: inline-block;
    top: 50%;
    left: 50%;
    width: ${checkbox.icon.width};
    height: ${checkbox.icon.height};
    transform: translate(-50%, -50%);
    pointer-events: none;

    & > svg {
      vertical-align: top;
    }
  `}
`;
IconWrapper.defaultProps = defaultProps;

const Icon = styled(({ className }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 10 8"
    className={className}
  >
    <path
      fillRule="evenodd"
      d="M.34 4.72a.485.485 0 0 1 0-.682l.68-.683a.478.478 0 0 1 .678 0L3.8 5.47 8.302.941a.478.478 0 0 1 .679 0l.678.683a.485.485 0 0 1 0 .683L4.14 7.859a.478.478 0 0 1-.678 0L.34 4.72z"
    />
  </svg>
))`
  ${({ theme: { checkbox } }) => css`
    width: ${checkbox.icon.width};
    height: ${checkbox.icon.height};

    & > path {
      fill: ${checkbox.icon.color};
    }
  `}
`;
Icon.defaultProps = defaultProps;

const Label = styled.label`
  cursor: pointer;
  vertical-align: middle;
`;

const Checkbox = forwardRef<HTMLInputElement, Props>(
  ({ children, className, ...rest }, ref) => (
    <Label>
      <CheckboxWrapper className={className}>
        <Input type="checkbox" ref={ref} {...rest} />
        <Box />
        <IconWrapper>
          <Icon />
        </IconWrapper>
      </CheckboxWrapper>

      {children}
    </Label>
  )
);

Checkbox.displayName = "Checkbox";

export { Checkbox };
