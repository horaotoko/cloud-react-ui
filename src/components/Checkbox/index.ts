export { CheckboxContainer as Checkbox } from "./CheckboxContainer";
export type { CheckboxProps } from "./CheckboxContainer";
