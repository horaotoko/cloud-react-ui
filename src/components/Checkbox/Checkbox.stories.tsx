import styled from "styled-components";
import { text, boolean } from "@storybook/addon-knobs";
import { Checkbox as CheckboxComponent } from "./Checkbox";
import { Text as TextComponent } from "../../components";

export default {
  component: CheckboxComponent,
  title: "components/Checkbox",
};

const Text = styled(TextComponent)`
  margin-left: 4px;
  margin-right: 10px;
`;

export const Checkbox = (): JSX.Element => {
  const disabled = boolean("Disabled", false);
  const label = text("Text", "Checkbox");

  return (
    <>
      <CheckboxComponent name="checkbox" disabled={disabled}>
        <Text>{label}</Text>
      </CheckboxComponent>
      <CheckboxComponent name="checkbox" disabled={disabled}>
        <Text>{label}</Text>
      </CheckboxComponent>
    </>
  );
};
