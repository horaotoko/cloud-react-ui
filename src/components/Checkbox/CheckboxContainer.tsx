import { CheckboxPresenter } from "./CheckboxPresenter";

export type CheckboxProps = {
  name?: string;
  checked?: boolean;
  disabled?: boolean;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick?: (event: React.MouseEvent<HTMLInputElement, MouseEvent>) => void;
  children?: React.ReactNode;
  className?: string;
};

export function CheckboxContainer(props: CheckboxProps): React.ReactElement {
  return <CheckboxPresenter {...props} />;
}
