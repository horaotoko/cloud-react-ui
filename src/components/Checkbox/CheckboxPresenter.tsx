import styled, { css } from "styled-components";
import { defaultProps } from "@theme";
import imageCheck from "@assets/images/check.svg";
import { CheckboxProps } from "./CheckboxContainer";

const StyledCheckbox = styled.input`
  ${({ theme }) => css`
    display: none;

    :checked + ${StyledText}:before {
      background-color: ${theme.checkbox.backgroundColor.checked};
      background-image: url(${imageCheck});
      background-size: ${theme.checkbox.backgroundSize};
      background-repeat: no-repeat;
      background-position: center;
      border-width: ${theme.checkbox.borderWidth.checked};
      border-style: ${theme.checkbox.borderStyle.checked};
      border-color: ${theme.checkbox.borderColor.checked};
    }

    :disabled + ${StyledText}:before {
      background-color: ${theme.checkbox.backgroundColor.disabled};
      border-width: ${theme.checkbox.borderWidth.disabled};
      border-style: ${theme.checkbox.borderStyle.disabled};
      border-color: ${theme.checkbox.borderColor.disabled};
    }
  `}
`;
StyledCheckbox.defaultProps = defaultProps;

const StyledText = styled.span`
  ${({ theme }) => css`
    position: relative;
    padding-left: 20px;
    font-size: ${theme.checkbox.fontSize};

    :before {
      content: "";
      box-sizing: border-box;
      display: block;
      position: absolute;
      top: 50%;
      left: 0;
      transform: translateY(-50%);
      width: ${theme.checkbox.width};
      height: ${theme.checkbox.height};
      background-color: ${theme.checkbox.backgroundColor.unchecked};
      border-width: ${theme.checkbox.borderWidth.unchecked};
      border-style: ${theme.checkbox.borderStyle.unchecked};
      border-color: ${theme.checkbox.borderColor.unchecked};
      border-radius: ${theme.checkbox.borderRadius};
  `}
`;
StyledText.defaultProps = defaultProps;

export function CheckboxPresenter({
  children,
  className,
  ...rest
}: CheckboxProps): React.ReactElement {
  return (
    <label className={className}>
      <StyledCheckbox type="checkbox" {...rest} />
      <StyledText>{children}</StyledText>
    </label>
  );
}
