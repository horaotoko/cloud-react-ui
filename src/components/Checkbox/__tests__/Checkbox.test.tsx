import { render, fireEvent, screen } from '@testing-library/react';
import { useState } from 'react';
import { Checkbox } from '../Checkbox';

describe('Checkbox', () => {
  it('default', () => {
    const { asFragment } = render(<Checkbox>checkbox</Checkbox>);
    expect(asFragment()).toMatchSnapshot();

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).not.toBeChecked();
    expect(checkbox).toBeEnabled();
  });

  it('defaultChecked props', () => {
    render(<Checkbox defaultChecked={true}>checkbox</Checkbox>);
    expect(screen.getByRole('checkbox')).toBeChecked();
  });

  it('disabled props', () => {
    render(<Checkbox disabled={true}>checkbox</Checkbox>);
    expect(screen.getByRole('checkbox')).toBeDisabled();
  });

  it('onChange, onClick props', () => {
    const onChange = jest.fn();
    const onClick = jest.fn();
    render(
      <Checkbox onChange={onChange} onClick={onClick}>
        checkbox
      </Checkbox>
    );
    fireEvent.click(screen.getByText('checkbox'));

    expect(onChange).toBeCalledTimes(1);
    expect(onClick).toBeCalledTimes(1);
  });

  it('checked, onChange props', () => {
    const TestForm = () => {
      const [checked, setChecked] = useState(false);
      const handleChange = () => {
        setChecked((prev) => !prev);
      };

      return (
        <Checkbox checked={checked} onChange={handleChange}>
          checkbox
        </Checkbox>
      );
    };

    render(<TestForm />);

    expect(screen.getByRole('checkbox')).not.toBeChecked();
    fireEvent.click(screen.getByText('checkbox'));
    expect(screen.getByRole('checkbox')).toBeChecked();
  });
});
