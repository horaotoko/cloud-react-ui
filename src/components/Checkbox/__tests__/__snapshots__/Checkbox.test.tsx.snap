// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox default 1`] = `
<DocumentFragment>
  .c1 {
  position: relative;
  display: inline-block;
  width: 16px;
  height: 16px;
  vertical-align: middle;
}

.c2 {
  opacity: 0;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  margin: 0;
  cursor: pointer;
}

.c2:checked + .c3 {
  background-color: #3B7DE9;
  border-width: 1px;
  border-style: solid;
  border-color: rgba(0,0,0,0.15);
}

.c2:disabled + .c3 {
  background-color: #D4D8DD;
  border-width: 1px;
  border-style: solid;
  border-color: #D4D8DD;
}

.c4 {
  position: absolute;
  width: 100%;
  height: 100%;
  border-radius: 4px;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #D4D8DD;
  pointer-events: none;
  box-sizing: border-box;
}

.c5 {
  position: absolute;
  display: inline-block;
  top: 50%;
  left: 50%;
  width: 10px;
  height: 8px;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  pointer-events: none;
}

.c5 > svg {
  vertical-align: top;
}

.c6 {
  width: 10px;
  height: 8px;
}

.c6 > path {
  fill: #FFFFFF;
}

.c0 {
  cursor: pointer;
  vertical-align: middle;
}

<label
    class="c0"
  >
    <span
      class="c1"
    >
      <input
        class="c2"
        type="checkbox"
      />
      <span
        class="c3 c4"
      />
      <span
        class="c5"
      >
        <svg
          class="c6"
          viewBox="0 0 10 8"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M.34 4.72a.485.485 0 0 1 0-.682l.68-.683a.478.478 0 0 1 .678 0L3.8 5.47 8.302.941a.478.478 0 0 1 .679 0l.678.683a.485.485 0 0 1 0 .683L4.14 7.859a.478.478 0 0 1-.678 0L.34 4.72z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </span>
    checkbox
  </label>
</DocumentFragment>
`;
